### ISSUE TITLE
Have you reproduced the accuracy of CircleLoss?

### ISSUE THREAD

**Issue by Kewenjing1020** [JDAI-CV/fast-reid#41]
I tried to use the implemented circleloss, but failed to reproduce the claimed accuracy in paper.

**Comment by L1aoXingyu** [comment]
you need to tune batch size with K=4, P=4, which means batch size is 16, with 4 identities and 4 images of each identity. Besides, it needs to freeze backbone for the first 2000 iters.
You can try on market151, I will put the `config.yaml` this weekend.

**Comment by L1aoXingyu** [comment]
![image](https://user-images.githubusercontent.com/19245504/80170815-6bdb6200-861b-11ea-9a76-993beb5d7a2f.png)
@Kewenjing1020 this is my reproduced results with config

```
MODEL:
  META_ARCHITECTURE: 'Baseline'

  OPEN_LAYERS: "heads"

  BACKBONE:
    NAME: "build_resnet_backbone"
    DEPTH: 50
    LAST_STRIDE: 1
    WITH_IBN: False
    WITH_NL: False
    PRETRAIN: True

  HEADS:
    NUM_CLASSES: 751
    NAME: "BNneckHead"
    CLS_LAYER: "circle"
    POOL_LAYER: "avgpool"
    SCALE: 128
    MARGIN: 0.25

  LOSSES:
    NAME: ("CrossEntropyLoss",)
    CE:
      EPSILON: 0.1
      SCALE: 1.0

DATASETS:
  NAMES: ("Market1501",)
  TESTS: ("Market1501",)

INPUT:
  SIZE_TRAIN: [256, 128]
  SIZE_TEST: [256, 128]
  REA:
    ENABLED: True
    PROB: 0.5
    MEAN: [123.675, 116.28, 103.53]
  DO_PAD: True


DATALOADER:
  PK_SAMPLER: True
  NUM_INSTANCE: 4
  NUM_WORKERS: 16

SOLVER:
  OPT: "SGD"
  MAX_ITER: 24000
  BASE_LR: 0.01
  BIAS_LR_FACTOR: 2.
  WEIGHT_DECAY: 0.0005
  WEIGHT_DECAY_BIAS: 0.0
  IMS_PER_BATCH: 16

  STEPS: [16000, 20000]
  GAMMA: 0.1

  WARMUP_FACTOR: 0.01
  WARMUP_ITERS: 4000
  FREEZE_ITERS: 4000

  LOG_PERIOD: 200
  CHECKPOINT_PERIOD: 6000

TEST:
  EVAL_PERIOD: 4000
  IMS_PER_BATCH: 512

CUDNN_BENCHMARK: True

OUTPUT_DIR: "logs/market/circle_loss_sgd"

```

you can reproduce my results in `BagTricks` folder by
```
CUDA_VISIBLE_DEVICES=1 python train_net.py --config-file='circle_loss.yml'
```

**Comment by Chaucergit** [comment]
StrongerBaseline项目中的配置文件时已经包含了circle loss的配置吗？我看使用的loss依然是NAME: ("CrossEntropyLoss", "TripletLoss")，没有配置loss为CircleLoss，当我改为CircleLoss是出现AttributeError: SCALE_TRI，是因为什么呢？

**Comment by L1aoXingyu** [comment]
@Chaucergit 不好意思，现在项目还处于快速迭代阶段，还没有合适的文档教程，circle loss 是通过 head 部分的分类器选择 `MODEL.HEADS.CLS_LAYER = 'circle'` 来实现 class-level based circle loss，类似于 arcface。

**Comment by Chaucergit** [comment]
好的，谢谢啦，我在继续看看你们的项目，在这里还是要给你们点赞的

**Comment by gepu0221** [comment]
> ![image](https://user-images.githubusercontent.com/19245504/80170815-6bdb6200-861b-11ea-9a76-993beb5d7a2f.png)
> @Kewenjing1020 this is my reproduced results with config
> 
> ```
> MODEL:
>   META_ARCHITECTURE: 'Baseline'
> 
>   OPEN_LAYERS: "heads"
> 
>   BACKBONE:
>     NAME: "build_resnet_backbone"
>     DEPTH: 50
>     LAST_STRIDE: 1
>     WITH_IBN: False
>     WITH_NL: False
>     PRETRAIN: True
> 
>   HEADS:
>     NUM_CLASSES: 751
>     NAME: "BNneckHead"
>     CLS_LAYER: "circle"
>     POOL_LAYER: "avgpool"
>     SCALE: 128
>     MARGIN: 0.25
> 
>   LOSSES:
>     NAME: ("CrossEntropyLoss",)
>     CE:
>       EPSILON: 0.1
>       SCALE: 1.0
> 
> DATASETS:
>   NAMES: ("Market1501",)
>   TESTS: ("Market1501",)
> 
> INPUT:
>   SIZE_TRAIN: [256, 128]
>   SIZE_TEST: [256, 128]
>   REA:
>     ENABLED: True
>     PROB: 0.5
>     MEAN: [123.675, 116.28, 103.53]
>   DO_PAD: True
> 
> 
> DATALOADER:
>   PK_SAMPLER: True
>   NUM_INSTANCE: 4
>   NUM_WORKERS: 16
> 
> SOLVER:
>   OPT: "SGD"
>   MAX_ITER: 24000
>   BASE_LR: 0.01
>   BIAS_LR_FACTOR: 2.
>   WEIGHT_DECAY: 0.0005
>   WEIGHT_DECAY_BIAS: 0.0
>   IMS_PER_BATCH: 16
> 
>   STEPS: [16000, 20000]
>   GAMMA: 0.1
> 
>   WARMUP_FACTOR: 0.01
>   WARMUP_ITERS: 4000
>   FREEZE_ITERS: 4000
> 
>   LOG_PERIOD: 200
>   CHECKPOINT_PERIOD: 6000
> 
> TEST:
>   EVAL_PERIOD: 4000
>   IMS_PER_BATCH: 512
> 
> CUDNN_BENCHMARK: True
> 
> OUTPUT_DIR: "logs/market/circle_loss_sgd"
> ```
> 
> you can reproduce my results in `BagTricks` folder by
> 
> ```
> CUDA_VISIBLE_DEVICES=1 python train_net.py --config-file='circle_loss.yml'
> ```

请问有在类别更多的数据集上训练过吗，我看这个类别数只有751？我在人脸识别的数据集上使用 circleloss，类别数为1000时收敛较快，类别数为1w 及以上，loss 收敛的特别慢，想问问你们有没有在类别数较大的数据集上训练过。

**Comment by L1aoXingyu** [comment]
@gepu0221 这两天我会试试在 3w id 的数据集上训练，看看效果

**Comment by my462** [comment]
您好，根据您的配置文件，您是只使用了circleloss的，但是根据您贴出来的结果图片，又是有tripletloss的，请问您复现结果的时候使用的是circleloss和tripletloss联合学习吗

**Comment by L1aoXingyu** [comment]
@my462 我记得之前跑过只有 circle loss 的结果，基本和论文差不多，加上 triplet loss，还能更高，可以参考 [model zoo](https://github.com/JDAI-CV/fast-reid/blob/master/MODEL_ZOO.md) 里面 sbs 的结果，这些结果都是 circle loss 和 triplet loss 联合 train 出来的

**Comment by my462** [comment]
@L1aoXingyu 对的，我已经用Circle loss替换了softmax，保留triplet loss在训练了，从原理上看这么训练应该能得到更好的结果；我也在用另一种论文里没有使用的方式，替换triplet loss为pair wise的circleloss，保留softmax loss训练中。目前看来前者scale的超参比较适合大一点的数值，如论文里的256，前者的方式scale需要设置的小一些，如32/64。请问您有做过这种方式的尝试吗？

**Comment by L1aoXingyu** [comment]
@my462 这个我没有尝试过，你可以试试，不过根据人脸的经验，ID 规模比较大之后，把 softmax 换成 circle loss 应该比较好，或者可以试试把 softmax 和 triplet loss 都换成 circle loss，然后调一下 scale，期待你的实验结果

**Comment by L1aoXingyu** [comment]
@gepu0221 最近试了一下 3w 的 id 规模上训练，`s=128`，loss 从 100 左右开始下降，2w iter 的时候是 52，5w iter 是 27，10w iter 是 20 左右，可以不用看 loss，直接看测试结果更好
BTW，我们在 evaluation 里面加入了 ROC 曲线和 TPR@FPR 这个指标
![roc](https://user-images.githubusercontent.com/19245504/82411285-1a9c8200-9aa4-11ea-86d8-dff70bc925c6.jpg)
![fpr](https://user-images.githubusercontent.com/19245504/82411385-43247c00-9aa4-11ea-8089-6e320e916c17.jpg)

**Comment by xiaodaodao123** [comment]
@my462 您好,请问您在reid中使用circle loss训练做什么特别的处理了吗,还是直接代替了softmax,可以不可以看一下您的实现呢

**Comment by ahzz1207** [comment]
我想问一下你们使用circleloss网络梯度咋样？ 我在文本分类上，随着loss的从70下降到4，网络的梯度值依然在剧烈震荡，且数值非常大（是使用softmax的10到100倍），请问你们在reid任务上有相似情况吗？

**Comment by L1aoXingyu** [comment]
@ahzz1207 我还没有关注过 circle loss 的梯度问题，不过 loss 确实会比较大，centerloss 和 circle loss 不冲突的，可以一起使用。

**Comment by ahzz1207** [comment]
> @ahzz1207 我还没有关注过 circle loss 的梯度问题，不过 loss 确实会比较大，centerloss 和 circle loss 不冲突的，可以一起使用。
我是这么考虑的， circleloss用的是单位化后的余弦相似度， centerloss用的是欧式距离， 是否应该把centerloss里的center先单位化后再求欧式距离呢？（此时和余弦本质上一致）。

**Comment by ivylinden** [comment]
CUDA_VISIBLE_DEVICES=1 python train_net.py --config-file='circle_loss.yml'  怎么找不到circle_loss.yaml  训练MGN+circleloss  MAX_ITER: 60是吗

**Comment by TCHeish** [comment]
> 我想问一下你们使用circleloss网络梯度咋样？ 我在文本分类上，随着loss的从70下降到4，网络的梯度值依然在剧烈震荡，且数值非常大（是使用softmax的10到100倍），请问你们在reid任务上有相似情况吗？
> 模型的acc是和softmax差不多的甚至高一点，同任务上centerloss也会梯度震荡，但是也只有circle的1/10大小，且似乎性能稍好些。

HI, 

Can you share the code of how you visualize the gradient? 
I guess you're using tensorborad, right? 
But I'm not sure how to apply in this(fastreid) libs.

Thanks!

**Comment by L1aoXingyu** [comment]
> CUDA_VISIBLE_DEVICES=1 python train_net.py --config-file='circle_loss.yml' 怎么找不到circle_loss.yaml 训练MGN+circleloss MAX_ITER: 60 OPT：Adam 是吗 steps是多少

mgn 训练 circle loss 就是下面这个 config

https://github.com/JDAI-CV/fast-reid/blob/master/configs/Base-MGN.yml

**Comment by wj1tr0y** [comment]
@L1aoXingyu 您好，非常感谢您的工作。有个问题我想问一下，circle loss的目的是优化sp->1，sn->0，最终的决策边界是sn-sp=m。那么按照我的理解sp应该趋近于1，sn趋近于0，但是在训练的时候我发现sp和sn会同时下降，并且间隔大概在0.03左右，请问这是为什么呢？
我在做reid任务，m=0.25，gamma=128，person=4， image=8。如果我用triplet预训练的模型加上circle loss，sp和sn就拉开的很大，但是从0开始训circle loss， sp和sn会共同下降，并且数值很相近。

**Comment by L1aoXingyu** [comment]
@wj1tr0y 那最后用 triplet 预训练的模型+circle loss 和从零训的 circle loss 比如何呢？

你的这个问题可能需要孙博来回答 @YifanSun-ReID

**Comment by wj1tr0y** [comment]
> @wj1tr0y 那最后用 triplet 预训练的模型+circle loss 和从零训的 circle loss 比如何呢？
> 
> 你的这个问题可能需要孙博来回答 @YifanSun-ReID

用triplet预训练的模型开始的时候sp->0.85 sn->0.30 训练到最后sp->0.79 sn->0.14 这个看起来是很make sense的。
但是从零训练circle loss，不仅circle loss不太收敛，并且sp和sn都趋近于相同。
这个现象从 [这个mnist example](https://github.com/TinyZeaMays/CircleLoss/issues/27)中也可以观察到。

**Comment by L1aoXingyu** [comment]
> 用triplet预训练的模型开始的时候sp->0.85 sn->0.30 训练到最后sp->0.79 sn->0.14 这个看起来是很make sense的。
> 但是从零训练circle loss，不仅circle loss不太收敛，并且sp和sn都趋近于相同。
> 这个现象从 [这个mnist example](https://github.com/TinyZeaMays/CircleLoss/issues/27)中也可以观察到。

那不用 triplet 预训练的模型，直接 train circle loss，完全无法收敛吗？那 triplet + circle 联合训练呢？

**Comment by wj1tr0y** [comment]
> > 用triplet预训练的模型开始的时候sp->0.85 sn->0.30 训练到最后sp->0.79 sn->0.14 这个看起来是很make sense的。
> > 但是从零训练circle loss，不仅circle loss不太收敛，并且sp和sn都趋近于相同。
> > 这个现象从 [这个mnist example](https://github.com/TinyZeaMays/CircleLoss/issues/27)中也可以观察到。
> 
> 那不用 triplet 预训练的模型，直接 train circle loss，完全无法收敛吗？那 triplet + circle 联合训练呢？

是的 直接train circle loss m=0.25，gamma128 训练15w个iter，circle loss大概在70左右，测试集acc也很低。联合训练也无法收敛。

**Comment by L1aoXingyu** [comment]
@wj1tr0y 你用的是 pair-wise version 的 circle loss 吧？

**Comment by wj1tr0y** [comment]
> @wj1tr0y 你用的是 pair-wise version 的 circle loss 吧？

是的 我也用了sparse circle loss，效果要好一点，但是也达不到triplet loss的精度。但是从特征角度来说的话，直接用pair-wise计算距离也应该是合理的呀

**Comment by L1aoXingyu** [comment]
@wj1tr0y 我这边用 pair-wise circle loss 也不太 work，所以我现在都是 softmax_circle loss + triplet loss，一直在等孙博 release code，不过之前我听他说最后 pair-wise circle 的 loss 就是很大。

**Comment by wj1tr0y** [comment]
@L1aoXingyu 好滴，那只能等孙博release了看看。非常感谢！！

**Comment by GuHongyang** [comment]
@L1aoXingyu 请问用circle loss训练MGN时，是像原文中说的那样将MGN的8个特征拼接在一起用一个circle loss训练吗？我邮件孙博了，他用的和原始的MGN的结构还不太一样，请问直接用MGN+circle loss能达到多少的精度呢？我这边尝试的结果为mAP=84.09%,Rank-1=93.94%。最终loss下降到2.5左右。所采用的参数为(gamma=256.0, m=0.25). 请问大佬能做到多少精度？

**Comment by chernbo** [comment]
@L1aoXingyu  感谢您和您团队的出色工作，我看了您提供的配置文件后，有两个疑惑：1、MAX_ITER：24000，在我的机器显示要跑18天，应该是这里版本和现在版本不同导致MAX_ITER意义不同导致的，2、我看到您提供的截图有Triplet loss这一项，但是您提供的配置文件没有使用到Triplet loss。这里似乎有一些问题。

**Comment by chernbo** [comment]
跟您给定配置文件，我自己尝试一下在bagtricks发现可以达到r1 94.69%,mAP:86.61%的效果
